drivers/mbox/mbox_stm32_hsem.c

```diff
@@ -0,0 +1,252 @@
+static inline void stm32_hsem_disable_rx_interrupt(void)
+{
+	const uint32_t mask_hsem_id = (1U << MBOX_RX_HSEM_ID);
+
+#if HSEM_CPU_ID == HSEM_CPU1
+	LL_HSEM_DisableIT_C1IER(HSEM, mask_hsem_id);
+#else /* HSEM_CPU2 */
+	LL_HSEM_DisableIT_C2IER(HSEM, mask_hsem_id);
+#endif /* HSEM_CPU_ID */
+}
+
+static inline void stm32_hsem_clear_rx_interrupt(void)
+{
+	const uint32_t mask_hsem_id = (1U << MBOX_RX_HSEM_ID);
+
+#if HSEM_CPU_ID == HSEM_CPU1
+	LL_HSEM_ClearFlag_C1ICR(HSEM, mask_hsem_id);
+#else /* HSEM_CPU2 */
+	LL_HSEM_ClearFlag_C2ICR(HSEM, mask_hsem_id);
+#endif /* HSEM_CPU_ID */
+}
+
+static inline uint32_t stm32_hsem_is_rx_interrupt_active(void)
+{
+	const uint32_t mask_hsem_id = (1U << MBOX_RX_HSEM_ID);
+
+#if HSEM_CPU_ID == HSEM_CPU1
+	return LL_HSEM_IsActiveFlag_C1ISR(HSEM, mask_hsem_id);
+#else /* HSEM_CPU2 */
+	return LL_HSEM_IsActiveFlag_C2ISR(HSEM, mask_hsem_id);
+#endif /* HSEM_CPU_ID */
+}
+
+static inline bool is_rx_channel_valid(const struct device *dev, uint32_t ch)
+{
+	/* Only support one RX channel */
+	return (ch == MBOX_RX_CHANNEL);
+}
+
+static inline bool is_tx_channel_valid(const struct device *dev, uint32_t ch)
+{
+	/* Only support one TX channel */
+	return (ch == MBOX_TX_CHANNEL);
+}
+
+static void mbox_dispatcher(const struct device *dev)
+{
+	struct mbox_stm32_hsem_data *data = dev->data;
+
+	/* Check semaphore rx_semid interrupt status */
+	if (!stm32_hsem_is_rx_interrupt_active())
+		return;
+
+	if (data->cb != NULL) {
+		data->cb(dev, MBOX_RX_CHANNEL, data->user_data, NULL);
+	}
+
+	/* Clear semaphore rx_semid interrupt status and masked status */
+	stm32_hsem_clear_rx_interrupt();
+}
+
+static int mbox_stm32_hsem_send(const struct device *dev, uint32_t channel,
+			 const struct mbox_msg *msg)
+{
+	if (msg) {
+		LOG_ERR("Sending data not supported.");
+		return -EINVAL;
+	}
+
+	if (!is_tx_channel_valid(dev, channel)) {
+		return -EINVAL;
+	}
+
+	/*
+	 * Locking and unlocking the hardware semaphore
+	 * causes an interrupt on the receiving side.
+	 */
+	z_stm32_hsem_lock(MBOX_TX_HSEM_ID, HSEM_LOCK_DEFAULT_RETRY);
+	z_stm32_hsem_unlock(MBOX_TX_HSEM_ID);
+
+	return 0;
+}
+
+static int mbox_stm32_hsem_register_callback(const struct device *dev, uint32_t channel,
+				      mbox_callback_t cb, void *user_data)
+{
+	struct mbox_stm32_hsem_data *data = dev->data;
+
+	if (!(is_rx_channel_valid(dev, channel))) {
+		return -EINVAL;
+	}
+
+	data->cb = cb;
+	data->user_data = user_data;
+
+	return 0;
+}
+
+static int mbox_stm32_hsem_mtu_get(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	/* We only support signalling */
+	return 0;
+}
+
+static uint32_t mbox_stm32_hsem_max_channels_get(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	/* Only two channels supported, one RX and one TX */
+	return 2;
+}
+
+static int mbox_stm32_hsem_set_enabled(const struct device *dev, uint32_t channel, bool enable)
+{
+	if (!is_rx_channel_valid(dev, channel)) {
+		return -EINVAL;
+	}
+
+	if (enable) {
+		stm32_hsem_clear_rx_interrupt();
```

Comment:
I agree, that is problematic. I have moved the clearing over to the initialization. Correct me if I am wrong, but since we have two separate registers for the interrupt active bit (one for each core), we need to clean it from both cores. I have implemented it this way.

Reply:
Resolving this now.